components/raftstore/src/store/worker/split_controller.rs

```diff
@@ -0,0 +1,532 @@
+            }
+        }
+        Recorder::split_key(
+            samples,
+            config.split_balance_score,
+            config.split_contained_score,
+            config.sample_threshold,
+        )
+    }
+
+    fn sample(samples: &mut Vec<Sample>, key_range: &KeyRange) {
+        for mut sample in samples.iter_mut() {
+            let order_start = if key_range.start_key.is_empty() {
+                Ordering::Greater
+            } else {
+                sample.key.cmp(&key_range.start_key)
+            };
+
+            let order_end = if key_range.end_key.is_empty() {
+                Ordering::Less
+            } else {
+                sample.key.cmp(&key_range.end_key)
+            };
+
+            if order_start == Ordering::Greater && order_end == Ordering::Less {
+                sample.contained += 1;
+            } else {
+                if order_start != Ordering::Greater {
+                    sample.right += 1;
+                } else if order_start == Ordering::Greater {
+                    sample.left += 1;
+                }
+            }
+        }
+    }
+
+    fn split_key(
+        samples: Vec<Sample>,
+        split_balance_score: f64,
+        split_contained_score: f64,
+        sample_threshold: i32,
+    ) -> Vec<u8> {
+        let mut best_index: i32 = -1;
+        let mut best_score = 2.0;
+        for index in 0..samples.len() {
+            let sample = &samples[index];
+            let sampled = sample.contained + sample.left + sample.right;
+            if (sample.left + sample.right) == 0 || sampled < sample_threshold {
+                continue;
+            }
+            let diff = (sample.left - sample.right) as f64;
+            let balance_score = diff.abs() / (sample.left + sample.right) as f64;
+            if balance_score >= split_balance_score {
+                continue;
+            }
+            let contained_score = sample.contained as f64 / sampled as f64;
+            if contained_score >= split_contained_score {
+                continue;
+            }
+            let final_score = balance_score + contained_score;
+            if final_score < best_score {
+                best_index = index as i32;
+                best_score = final_score;
+            }
+        }
+        if best_index >= 0 {
+            return samples[best_index as usize].key.clone();
+        }
+        return vec![];
+    }
+}
+
+#[derive(Clone)]
+pub struct QpsStats {
+    pub region_infos: HashMap<u64, RegionInfo>,
+    pub sample_num: usize,
+}
+
+impl QpsStats {
+    pub fn new() -> QpsStats {
+        QpsStats {
+            sample_num: DEFAULT_SAMPLE_NUM,
+            region_infos: HashMap::default(),
+        }
+    }
+
+    pub fn add(&mut self, region_id: u64, peer: &Peer, key_range: KeyRange) {
+        self.batch_add(region_id, peer, vec![key_range]);
+    }
+
+    pub fn batch_add(&mut self, region_id: u64, peer: &Peer, key_ranges: Vec<KeyRange>) {
```

Comment:
Do we need both `add` and `batch_add`?

Reply:
they will be used to `get` or `batch_get`